resnet/main.py

```diff
@@ -0,0 +1,400 @@
+"""
+    PyTorch training code for Wide Residual Networks:
+    http://arxiv.org/abs/1605.07146
+
+    The code reproduces *exactly* it's lua version:
+    https://github.com/szagoruyko/wide-residual-networks
+
+    2016 Sergey Zagoruyko
+"""
+
+import argparse
+import os
+import json
+import numpy as np
+import math
+from tqdm import tqdm
+from nested_dict import nested_dict
+from collections import OrderedDict
+import torch
+import torch.optim
+import torchvision.datasets as datasets
+from torch.autograd import Variable
+import torch.nn.functional as F
+from torch.legacy.nn import SpatialZeroPadding, SpatialReflectionPadding
+import torchnet as tnt
+from torchnet.engine import Engine
+import torchnet.meter as meter
+import torch.cuda.comm as comm
+from torch.nn.parallel._functions import Broadcast
+from torch.nn.parallel import scatter, parallel_apply, gather
+import torch.backends.cudnn as cudnn
+
+cudnn.benchmark = True
+
+parser = argparse.ArgumentParser(description='Wide Residual Networks')
+# Model options
+parser.add_argument('--model', default='resnet', type=str)
+parser.add_argument('--depth', default=16, type=int)
+parser.add_argument('--width', default=1, type=float)
+parser.add_argument('--dataset', default='CIFAR10', type=str)
+parser.add_argument('--dtype', default='float', type=str)
+parser.add_argument('--nthread', default=4, type=int)
+parser.add_argument('--dropout', default=0, type=float)
+
+# Training options
+parser.add_argument('--batchSize', default=128, type=int)
+parser.add_argument('--lr', default=0.1, type=float)
+parser.add_argument('--epochs', default=200, type=int, metavar='N',
+                    help='number of total epochs to run')
+parser.add_argument('--weightDecay', default=0.0005, type=float)
+parser.add_argument('--epoch_step', default='[60,120,160]', type=str,
+                    help='json list with epochs to drop lr on')
+parser.add_argument('--lr_decay_ratio', default=0.2, type=float)
+parser.add_argument('--resume', default='', type=str)
+
+# Device options
+parser.add_argument('--save', default='/tmp/cifar', type=str,
+                    help='save parameters and logs in this folder')
+parser.add_argument('--ngpu', default=1, type=int,
+                    help='number of GPUs to use for training, 0 to for CPU')
+parser.add_argument('--gpu_id', default='0', type=str,
+                    help='id(s) for CUDA_VISIBLE_DEVICES')
+
+# TODO: move this to main
+opt = parser.parse_args()
+print 'parsed options:', vars(opt)
+
+os.environ['CUDA_VISIBLE_DEVICES'] = opt.gpu_id
```

Comment:
You should never use `set_device` in regular scripts, we've added it only for iPython notebooks where it's sometimes convenient. `CUDA_VISIBLE_DEVICES` is better, because it will prevent any memory alocation on unused GPUs (every PyTorch process uses ~250MB memory on each GPU it sees).